Add Profile credentials providers

oharaandrew314 opened the pull request:
- Added a `CredentialsChain` which allows for the return-value to be null, permitting chaining operations
- `CredentialsChain.Profile` gets from a local credentials file 
- `CredentialsChain.Ec2InstanceProfile` gets from the EC2 Instance Metadata Service
- `CredentialsProvider.Default` will attempt to load credentials in the following order: `Environment`, `Profile`, `Ec2InstanceProfile`
- Migrated existing clients to `CredentialsProvider.Default`

oharaandrew314 commented:
Tests are passing now; just the automerge test that's failing.

daviddenton commented:
We're actually not at all fans of the credentials chain - so could you possibly at the very least remove the defaulting of it . We very specifically didn't implement the concept of the chain as we have had *really* bad experiences with not knowing where clients are getting their credentials and magic things happening - it's one of our big bug-bears with the AWS SDK.

I've fixed the automerge thing so at least PR builds shouldn't fail any more 😄

oharaandrew314 commented:
Fair enough; I've reverted the change to the default `CredentialsProvider`, and I've updated the nomenclature of the default chain.  I like the default chain from the SDK because it means I don't need to worry about where the code is being run; be it on a developer machine, on an EC2, or in a Lambda.  But at least I still have the option to use it.

I can't really restore the composite `AWS_CREDENTIALS` Environment Lens, because I still need the `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY` to be optional for the chained version... unless you want optional duplicates.

daviddenton commented:
>I can't really restore the composite AWS_CREDENTIALS Environment Lens, because I still need the AWS_ACCESS_KEY_ID and AWS_SECRET_ACCESS_KEY to be optional for the chained version... unless you want optional duplicates.

Can we please keep these as required and have the duplicates just in the chain (and have the AWS_CREDENTIALS lens restored)? We use these all over the place so it would be a really disruptive change 😄

oharaandrew314 commented:
No problem; I've restored it.

oharaandrew314 commented:
Actually, I just realized something.  I believe `CredentialsProvider` is invoked per request, which means I'm going to have to implement caching for the `Profile` and especially the `Ec2InstanceProfile` chains.  The latter will most likely require that I honour the expiration of the STS credentials.

daviddenton commented:
Ok - let me know when you think it's ready 😄
